tests/stub/authorization.py

```diff
@@ -655,3 +657,69 @@ def get_vars(self, host=None):
 
     def get_db(self):
         return None
+
+
+class NoRoutingAuthorizationTests(BaseAuthorizationTests):
+    def setUp(self):
+        super().setUp()
+        self._readServer = StubServer(9010)
+        self._uri = "bolt://%s:%d" % (self._readServer.host,
+                                      self._readServer.port)
+        self._auth = types.AuthorizationToken(
+            scheme="basic", principal="p", credentials="c")
+        self._userAgent = "007"
+
+    def tearDown(self):
+        self._readServer.reset()
+        super().tearDown()
+
+    def read_return_1_failure_return_2_succeed_script(self):
+        return """
+        !: BOLT 4
+        !: AUTO HELLO
+        !: AUTO GOODBYE
+        !: AUTO RESET
+        !: ALLOW CONCURRENT
+
+        {{
+            C: RUN "RETURN 1 as n" {} {"mode": "r"}
+            C: PULL {"n": 1000}
+            S: SUCCESS {"fields": ["n"]}
+               FAILURE {"code": "Neo.ClientError.Security.AuthorizationExpired", "message": "Authorization expired"}
+        ----
+            C: RUN "RETURN 2 as n" {} {"mode": "r"}
+            C: PULL {"n": 1000}
+            S: SUCCESS {"fields": ["n"]}
+               RECORD [1]
+               SUCCESS {"type": "r"}
+            C: RUN "RETURN 2 as n" {} {"mode": "r"}
+            C: PULL {"n": 1000}
+            S: SUCCESS {"fields": ["n"]}
+               RECORD [1]
+               SUCCESS {"type": "r"}
+        }}
+        """
+
+    def test_should_drop_connection_after_AuthorizationExpired(self):
+        self._readServer.start(
+            script=self.read_return_1_failure_return_2_succeed_script()
+        )
+        driver = Driver(self._backend, self._uri, self._auth,
+                        userAgent=self._userAgent)
+
+        session1 = driver.session('r')
+        session2 = driver.session('r')
+
+        session1.run('RETURN 2 as n').next()
+
+        try:
+            session2.run('RETURN 1 as n').next()
+        except types.DriverError as e:
+            self.assert_is_authorization_error(e)
+
+        session2.close()
+        session1.close()
+        accept_count = self._readServer.count_responses("<ACCEPT>")
+        hangup_count = self._readServer.count_responses("<HANGUP>")
+        
+        self.assertEqual(accept_count, hangup_count)
```

Comment:
This needs some refinement as it very likely to fail on the java driver, which does not hangup on all connections immediately, but makes sure that they do net get reused. In other words, the java driver performs a lazy cleanup on connection acquisition.

Reply:
So, maybe it could be the case load the accept count at this stage. Then create a new session (run a query to force the connection be fetched, JS stuff :D ) and then count fetch the HANGUPs to compare. 

I will try to do this refinement and run against the JS driver (with and without the fix)

Reply:
@injectives I updated the test with a more lenient implementation.

Reply:
The changes look better. 👍  I have actually implemented the feature list support in Java Driver: https://github.com/neo4j/neo4j-java-driver/pull/889
However, I am slightly concerned about the accept and hangup counts because I get either `1` or `2` for both. Obviously, the test passes in both cases but it seems like we should find an answer why it happens first.
On a separate note, does the new stub server always forcibly hangup when it returns the auth expired error? (this is would match the real server behaviour)

Reply:
It's supposed to always be at least two connections to hangup, since it is running two sessions in parallel. But i don't know if it have more than 2 connections open and then closed will be an issue for this test in particular. 

I will add the exit after the exit on the server failure, to make it closer to the real world behaviour.

Reply:
While it looks like it should run 2 sessions in parallel, this is not necessarily the case. For instance, the Java driver is slightly eager and tries looking ahead, which causes it to realise that the whole data has been consumed and there is nothing to do in `session1` and it releases the connection. This sometimes results in `session1` and `session2` effectively using the same connection and introduces an unexpected/unanticipated variation to the test.

This may apply to any driver that tries looking ahead. I managed to get it to work as expected with the following changes:
1. Update the script to support multiple pulls and expect `"n": 1`
```
{+
    {{
        C: RUN "RETURN 1 as n" {} {"mode": "r"}
        C: PULL {"n": 1000}
        S: SUCCESS {"fields": ["n"]}
           FAILURE {"code": "Neo.ClientError.Security.AuthorizationExpired", "message": "Authorization expired"}
        S: <EXIT>
    ----
        C: RUN "RETURN 2 as n" {} {"mode": "r"}
        S: SUCCESS {"fields": ["n"]}
        {+
            C: PULL {"n": 1}
            S: RECORD [1]
               SUCCESS {"type": "r", "has_more": true}
        +}
        C: DISCARD {"n": -1}
        S: SUCCESS {}
    ----
        C: RUN "RETURN 3 as n" {} {"mode": "r"}
        C: PULL {"n": 1000}
        S: SUCCESS {"fields": ["n"]}
           RECORD [1]
           SUCCESS {"type": "r"}
    }}

+}
```
2. Configure `session1` to have `fetchSize=1`:
```
session1 = driver.session('r', fetchSize=1)
```
3. Change query in `session3`:
```
session3.run('RETURN 3 as n').next()
```